File: 2022/src_py/d11.py
inspections = [0, 0, 0, 0, 0, 0, 0, 0]

# items = [[79, 98],
#          [54, 65, 75, 74],
#          [79, 60, 97],
#          [74]]

# lcm = 23 * 19 * 13 * 17

# new_item_fn = [lambda item: (item * 19) % lcm,
#                lambda item: (item + 6) % lcm,
#                lambda item: (item * item) % lcm,
#                lambda item: (item + 3) % lcm]

# new_monkey_fn = [lambda item: 2 if item % 23 == 0 else 3,
#                  lambda item: 2 if item % 19 == 0 else 0,
#                  lambda item: 1 if item % 13 == 0 else 3,
#                  lambda item: 0 if item % 17 == 0 else 1]

items = [[92, 73, 86, 83, 65, 51, 55, 93],
         [99, 67, 62, 61, 59, 98],
         [81, 89, 56, 61, 99],
         [97, 74, 68],
         [78, 73],
         [50],
         [95, 88, 53, 75],
         [50, 77, 98, 85, 94, 56, 89]]

lcm = 11 * 2 * 5 * 17 * 19 * 7 * 3 * 13

new_item_fn = [lambda item: (item * 5) % lcm,
               lambda item: (item * item) % lcm,
               lambda item: (item * 7) % lcm,
               lambda item: (item + 1) % lcm,
               lambda item: (item + 3) % lcm,
               lambda item: (item + 5) % lcm,
               lambda item: (item + 8) % lcm,
               lambda item: (item + 2) % lcm]

new_monkey_fn = [lambda item: 3 if item % 11 == 0 else 4,
                 lambda item: 6 if item % 2 == 0 else 7,
                 lambda item: 1 if item % 5 == 0 else 5,
                 lambda item: 2 if item % 17 == 0 else 5,
                 lambda item: 2 if item % 19 == 0 else 3,
                 lambda item: 1 if item % 7 == 0 else 6,
                 lambda item: 0 if item % 3 == 0 else 7,
                 lambda item: 4 if item % 13 == 0 else 0] 

monkeys = len(new_monkey_fn)

for round in range(10000):
    for monkey in range(monkeys):
        for item in items[monkey]:
            new_item = new_item_fn[monkey](item)
            new_monkey = new_monkey_fn[monkey](new_item)
            items[new_monkey].append(new_item)
            inspections[monkey] += 1
        items[monkey] = []

print(inspections)

File: 2022/src_old/d11.clj
(ns coa.d11)

; (def items
;   [[79, 98]
;    [54, 65, 75, 74]
;    [79, 60, 97]
;    [74]])

; (def monkeys
;   [{:op (fn [it] (quot (* it 19) 3)) :div 23 :t 2 :f 3}
;    {:op (fn [it] (quot (+ it 6) 3)) :div 19 :t 2 :f 0}
;    {:op (fn [it] (quot (* it it) 3)) :div 13 :t 1 :f 3}
;    {:op (fn [it] (quot (+ it 3) 3)) :div 17 :t 0 :f 1}])

(def items
  [[92, 73, 86, 83, 65, 51, 55, 93]
   [99, 67, 62, 61, 59, 98]
   [81, 89, 56, 61, 99]
   [97, 74, 68]
   [78, 73]
   [50]
   [95, 88, 53, 75]
   [50, 77, 98, 85, 94, 56, 89]])

(def monkeys
  [{:op (fn [it] (quot (* it 5) 3)) :div 11 :t 3 :f 4}
   {:op (fn [it] (quot (* it it) 3)) :div 2 :t 6 :f 7}
   {:op (fn [it] (quot (* it 7) 3)) :div 5 :t 1 :f 5}
   {:op (fn [it] (quot (+ it 1) 3)) :div 17 :t 2 :f 5}
   {:op (fn [it] (quot (+ it 3) 3)) :div 19 :t 2 :f 3}
   {:op (fn [it] (quot (+ it 5) 3)) :div 7 :t 1 :f 6}
   {:op (fn [it] (quot (+ it 8) 3)) :div 3 :t 0 :f 7}
   {:op (fn [it] (quot (+ it 2) 3)) :div 13 :t 4 :f 0}])

(def monkey-inspections (atom (vec (repeat (count monkeys) 0))))

(defn- register-inspection [inspections monkey-idx]
  (assoc inspections monkey-idx (inc (nth inspections monkey-idx))))

(defn- get-monkey-action [monkey-idx]
  (let [{:keys [op div t f]} (nth monkeys monkey-idx)]
    (fn [items item]
      (let [new-item (op item)
            next-monkey-idx (if (= (rem new-item div) 0) t f)
            next-monkey-items (conj (nth items next-monkey-idx) new-item)
            monkey-items (vec (rest (nth items monkey-idx)))]
        (swap! monkey-inspections register-inspection monkey-idx)
        (assoc items monkey-idx monkey-items next-monkey-idx next-monkey-items)
        ))))

(defn- monkey-turn [items monkey-idx]
  (let [monkey-items (nth items monkey-idx)
        monkey-action (get-monkey-action monkey-idx)]
    (reduce monkey-action items monkey-items)))

(defn- play-round [items _]
  (reduce monkey-turn items (range (count monkeys))))

(defn -main []
  (reduce play-round items (range 10000))
  (println (apply * (take 2 (reverse (sort @monkey-inspections))))))

(comment
  (-main))
